### PR TITLE
Initial work to move indentation services down to codestyle layer (step 8/N)

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
[Followup to https://github.com/dotnet/roslyn/pull/60445](https://github.com/dotnet/roslyn/pull/60448)

This PR moves the remainder of the IndentationService down.  The only thing remaining is moving existing that use this down to codestyle as well.

I'm doing these PRs in multiple parts to make things simpler to review. Every time i try to tackle the general problem, it explodes into a ton of changes.

The reason for this change overall is that we have fixers taht need to know indentation information. For example, the 'convert to file-scoped-namespace (and reverse)' fixes have to know indentation information so it can make the right fixes. Not having indentation info be available at thsi level means that the fixes are constrained to the VS host.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Formatting/Indentation/CSharpFormatterTestsBase.cs [hunk 1]
```diff
@@ -80,7 +81,7 @@ private static async Task TokenFormatWorkerAsync(TestWorkspace workspace, ITextB
 
             var formattingRuleProvider = workspace.Services.GetService<IHostDependentFormattingRuleFactoryService>();
 
-            var rules = formattingRuleProvider.CreateRule(document, position).Concat(Formatter.GetDefaultFormattingRules(document));
+            var rules = ImmutableArray.Create(formattingRuleProvider.CreateRule(document, position)).AddRange(Formatter.GetDefaultFormattingRules(document));
```
Comment (dibarbet) [inline src/EditorFeatures/CSharpTest/Formatting/Indentation/CSharpFormatterTestsBase.cs:84]: could the `IHostDependentFormattingRuleFactoryService` return an immutable array?  or does it need to be ienumerable for some consumer

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/AbstractIndentationService.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            // If the indenter can't produce a valid result, just default to 0 as our indentation.
+            return indenter.GetDesiredIndentation((FormattingOptions2.IndentStyle)indentStyle) ?? default;
+        }
+
+        private Indenter GetIndenter(Document document, int lineNumber, FormattingOptions2.IndentStyle indentStyle, CancellationToken cancellationToken)
+        {
+            var syntaxFormatting = this.SyntaxFormatting;
+
+#if CODE_STYLE
+            var tree = document.GetSyntaxTreeAsync(cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
```
Comment (dibarbet) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/AbstractIndentationService.cs:46]: to me would read better if there was just a `GetTreeAndIndentation` and `GetIndentationRule` method that kind of hid the codestyle messiness, but that might just be me

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/IIndentationService.cs [hunk 3]
```diff
@@ -73,7 +47,11 @@ internal static class IIndentationServiceExtensions
         /// </summary>
         public static string GetPreferredIndentation(this SyntaxToken token, Document document, CancellationToken cancellationToken)
         {
+#if CODE_STYLE
+            var sourceText = document.GetTextAsync(cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
```
Comment (dibarbet) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/IIndentationService.cs:51]: I thought the codestyle layer didn't use documents or am I totally off base

**Review by dibarbet — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/Formatting/Indentation/CSharpFormatterTestsBase.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Formatting/Indentation/CSharpFormatterTestsBase.cs:84]: teh host actually returns a single rule.  but then we need to wrap that in an immutable array to be able to addrange after that.  it's def not clean and i wish tehre was a better idiom for "i want to make an immutable array out of individual objects and streams of objects".

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/AbstractIndentationService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/AbstractIndentationService.cs:46]: yeah.  i don't love this. but i don't feel like there are really good alternaives here :(

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/IIndentationService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/IIndentationService.cs:51]: this is not the *codefixes* layer, which can use documents (but only the public facing side of them).

**Pull request merged**